src/components/mass-calculator/MassCalculator.test.tsx

```diff
@@ -0,0 +1,147 @@
+import MassCalculator from "./MassCalculator";
+import { Router } from "react-router-dom";
+import { createMemoryHistory } from "history";
+import { IElement } from "@/Element";
+import ElementManager, { getElementLocales } from "../../ElementManager";
+import "hammerjs";
+
+const elementsArr = [
+  {
+    atomic: 67,
+    symbol: "Ho",
+    name: "Holmium",
+    atomicMass: "164.93032",
+    cpkHexColor: "00FF9C",
+    electronicConfiguration: "[Xe] 4f11 6s2",
+    electronegativity: 1.23,
+    atomicRadius: 226,
+    ionRadius: "90.1 (+3)",
+    vanDelWaalsRadius: 2,
+    ionizationEnergy: 581,
+    electronAffinity: -50,
+    oxidationStates: "3",
+    standardState: "solid",
+    bondingType: "metallic",
+    meltingPoint: "1747",
+    boilingPoint: "2968",
+    density: 8.8,
+    group: "lanthanoids",
+    yearDiscovered: 1878,
+    valency: "+3",
+    wrongValences: ["+1", "+2", "+2 +3", "+3 +5", "+2 +4"],
+    testState: {
+      valencesTest: false,
+      ptTest: true,
+    },
+    realname: "abc",
+  },
+  {
+    atomic: 16,
+    symbol: "S",
+    name: "Sulfur",
+    atomicMass: "32.065",
+    cpkHexColor: "FFFF30",
+    electronicConfiguration: "[Ne] 3s2 3p4",
+    electronegativity: 2.58,
+    atomicRadius: 88,
+    ionRadius: "184 (-2)",
+    vanDelWaalsRadius: 180,
+    ionizationEnergy: 1000,
+    electronAffinity: -200,
+    oxidationStates: "-2, -1, 1, 2, 3, 4, 5, 6",
+    standardState: "solid",
+    bondingType: "covalent_network",
+    meltingPoint: "392.2",
+    boilingPoint: "717.82",
+    density: 1.96,
+    group: "amphigens",
+    yearDiscovered: 1878,
+    valency: "+2 +4 +6 / -2",
+    wrongValences: [
+      "+1 +3 +5 +7 / -1",
+      "+2 +4 +6",
+      "+3 +5 / -3",
+      "+3 +5",
+      "+2 +6",
+      "+2",
+    ],
+    testState: {
+      valencesTest: true,
+      ptTest: true,
+    },
+    realname: "abc",
+  },
+  {
+    atomic: 44,
+    symbol: "Ru",
+    name: "Ruthenium",
+    atomicMass: "101.07",
+    cpkHexColor: "248F8F",
+    electronicConfiguration: "[Kr] 4d7 5s1",
+    electronegativity: 2.2,
+    atomicRadius: 178,
+    ionRadius: "68 (+3)",
+    vanDelWaalsRadius: 188,
+    ionizationEnergy: 710,
+    electronAffinity: -101,
+    oxidationStates: "-2, 1, 2, 3, 4, 5, 6, 7, 8",
+    standardState: "solid",
+    bondingType: "metallic",
+    meltingPoint: "2610",
+    boilingPoint: "4425",
+    density: 12.37,
+    group: "transition_metals",
+    yearDiscovered: 1827,
+    valency: "+2 +3 +4 +6 +8",
+    wrongValences: [
+      "+2 +3 +4 +5 +6 +8",
+      "+2 +4 +6 +8",
+      "+1 +3 +5 +7 / -1",
+      "+1",
+      "+2",
+      "+2 +4",
+      "+2 +3",
+      "+1 +2",
+    ],
+    testState: {
+      valencesTest: false,
+      ptTest: true,
+    },
+    realname: "abc",
+  },
+];
+
+beforeEach(() => {
+  jest.clearAllMocks();
+  jest
+    .spyOn(ElementManager, "getElement")
+    .mockImplementation((atomic) =>
+      elementsArr.find((element) => element.atomic === atomic)
+    );
+});
```

Comment:
@mpeyper this is concerning to me, its a code implementation detail AND relies on the class/function having getter methods

Reply:
IF there isn't a better way I understand using this. 

@nobrayner  @Aprillion  @codyarose Thoughts?

Reply:
I'll glad to have some suggestions over this. As `ElementManager` is exposed as instance object with getter methods and doing my research I ended up spying over it.
Also that `hammer.js` gave me headache as in how to inform jest about Hammer. But finally that one's resolved.

Reply:
> I'll glad to have some suggestions over this. As `ElementManager` is exposed as instance object with getter methods and doing my research I ended up spying over it.
> Also that `hammer.js` gave me headache as in how to inform jest about Hammer. But finally that one's resolved.

These tests are also for refactoring to Hooks, so making tests around Class-based models is not ideal either.

Reply:
from the looks of the code, it gets the Element 
```js
 private massCalculatorElement(massCalculatorElement: IMassCalculatorElement) {
    const { atomic, quantity } = massCalculatorElement;
    const element = ElementManager.getElement(atomic); // Here is what you are mocking
```
`atomic` is part of the components initial state, which you can more reliably mock 
```
interface IMassCalculatorElement {
  atomic: number;
  quantity: number;
}

type Props = RouteComponentProps<any> & React.Props<any>;

interface IMassCalculatorState {
  elements: IMassCalculatorElement[];
  addElementModalOpen: boolean;
  modifyElementModal: {
    open: boolean;
    elementIndex: number;
  };
}
```

Reply:
Hope that helps

Reply:
It looks like `ElementManager` is a singleton. 

I think we just need to call the `ElementManager.loadElements();` at the very beginning of the test file in order to bootstrap the ElementManager with the necessary data.

Reply:
Actually I had a look around and it seems that the `ElementManager` is used by multiple components. 

Maybe we should just initiate the `ElementManager` and load all the elements within `setupTests`. 

E.G.

``` javascript
// src/setupTests.ts

import ElementManager from "./ElementManager";
ElementManager.loadElements();
```

From what I can see `ElementManager` does not modify any of the data that it exposes. `ElementManager` only provides getter functions to access the data so I shouldn't be an issue call `ElementManager.loadElements();` within `setupTests`?

Reply:
sounds like a good idea to `loadElements()` in setupTests.ts (unless we find some tests where we do NOT want to have it initialized of course)

if you do this, please don't forget update the test(s) that initialize it inside `beforeAll` - no need to do it twice.

Reply:
@JacobMGEvans - I would vote against mocking `ElementManager`, IMHO the real thing works fine in tests, e.g. https://github.com/OpenSourceRaidGuild/atom-pwa/blob/next/src/components/element-picker/ElementPicker.test.tsx

Reply:
Yeah, quite a few tests have been using the real version directly already without any issues. U suggest not mocking it and loading the elements up front is fine here.

Reply:
Thanks guys for suggestions. I'll add `ElementManager.loadElements()` inside setup file and will modify rest.

Reply:
Cool (Singleton! I knew I have seen that pattern before! lol) @grady-lad @Aprillion you can check out this comment too: https://github.com/OpenSourceRaidGuild/atom-pwa/pull/52#discussion_r584224172